### DISCUSSION TITLE
Chatacter None not found in the vocabulary

### DISCUSSION THREAD

**Discussion by OmiWakode** [coqui-ai/TTS#1489] [General Q&A]
While training Vits on my custom dataset( English) , i get this message printed in the terminal. What does it mean? Will this affect the model's performance and loss values?. 
```
[None, 't', None, 'h', None, 'e', None, ' ', None, 'b', None, 'i', None, 'r', None, 'c', None, 'h', None, ' ', None, 'c', None, 'a', None, 'n', None, 'o', None, 'e', None, ' ', None, 's', None, 'l', None, 'i', None, 'd', None, ' ', None, 'o', None, 'n', None, ' ', None, 't', None, 'h', None, 'e', None, ' ', None, 's', None, 'm', None, 'o', None, 'o', None, 't', None, 'h', None, ' ', None, 'p', None, 'l', None, 'a', None, 'n', None, 'k', None, 's', None]
 [!] Character None not found in the vocabulary. Discarding it.
```

Here is my config file
```
{
    "output_path": "/home/ubuntu/Custom_TTS_train/Result/",
    "logger_uri": null,
    "run_name": "vits_experiment",
    "project_name": "Voice cloning",
    "run_description": "",
    "print_step": 50,
    "plot_step": 100,
    "model_param_stats": false,
    "wandb_entity": null,
    "dashboard_logger": "tensorboard",
    "log_model_step": 10000,
    "save_step": 10000,
    "save_n_checkpoints": 2,
    "save_checkpoints": true,
    "save_all_best": false,
    "save_best_after": 10000,
    "target_loss": null,
    "print_eval": true,
    "test_delay_epochs": -1,
    "run_eval": true,
    "distributed_backend": "nccl",
    "distributed_url": "tcp://localhost:54321",
    "mixed_precision": false,
    "epochs": 5000,
    "batch_size": 16,
    "eval_batch_size": 16,
    "grad_clip": [
        5.0,
        5.0
    ],
    "scheduler_after_epoch": true,
    "lr": 0.00008,
    "optimizer": "AdamW",
    "optimizer_params": {
        "betas": [
            0.8,
            0.99
        ],
        "eps": 1e-09,
        "weight_decay": 0.01
    },
    "lr_scheduler": "StepLR",
    "lr_scheduler_params": {"step_size":1000, "verbose":true},
    "use_grad_scaler": false,
    "cudnn_enable": true,
    "cudnn_benchmark": true,
    "torch_seed": 54321,
    "model": "vits",
    "num_loader_workers": 4,
    "num_eval_loader_workers": 4,
    "use_noise_augment": false,
    "use_language_weighted_sampler": false,
    "audio": {
        "fft_size": 1024,
        "win_length": 1024,
        "hop_length": 256,
        "frame_shift_ms": null,
        "frame_length_ms": null,
        "stft_pad_mode": "reflect",
        "sample_rate": 16000,
        "resample": false,
        "preemphasis": 0.0,
        "ref_level_db": 20,
        "do_sound_norm": false,
        "log_func": "np.log",
        "do_trim_silence": true,
        "trim_db": 45,
        "do_rms_norm": false,
        "db_level": null,
        "power": 1.5,
        "griffin_lim_iters": 60,
        "num_mels": 80,
        "mel_fmin": 0.0,
        "mel_fmax": null,
        "spec_gain": 1,
        "do_amp_to_db_linear": false,
        "do_amp_to_db_mel": true,
        "pitch_fmax": 640.0,
        "pitch_fmin": 0.0,
        "signal_norm": false,
        "min_level_db": -100,
        "symmetric_norm": true,
        "max_norm": 4.0,
        "clip_norm": true,
        "stats_path": null
    },
    "use_phonemes": false,
    "phonemizer": null,
    "phoneme_language": "en-us",
    "compute_input_seq_cache": false,
    "text_cleaner": "multilingual_cleaners",
    "enable_eos_bos_chars": false,
    "test_sentences_file": "",
    "phoneme_cache_path": null,
    "characters": {
	"vocab_dict" : {"a":0, "b":1, "e":2, "f":7, "n": 10, "g": 8, "t": 11},
        "pad": "_",
        "eos": "&",
        "bos": "*",
        "blank": null,
        "characters": "ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz",
        "punctuations": "!'(),-.:;? ",
        "phonemes": "",
        "is_unique": true,
        "is_sorted": true
    },
    "add_blank": true,
    "batch_group_size": 0,
    "loss_masking": true,
    "sort_by_audio_len": false,
    "min_audio_len": 1,
    "max_audio_len": Infinity,
    "min_text_len": 1,
    "max_text_len": Infinity,
    "compute_f0": false,
    "compute_linear_spec": true,
    "precompute_num_workers": 0,
    "start_by_longest": false,
    "datasets": [
        {
            "name": "indian-speakers",
            "path": "/home/ubuntu/Custom_TTS_train/train_tts_approach/",
            "meta_file_train": "metadata_old.csv",
            "ignored_speakers": null,
            "language": "en",
            "meta_file_val": null,
            "meta_file_attn_mask": ""
        }
    ],
    "test_sentences": [
        [
            "It took me quite a long time to develop a voice, and now that I have it I'm not going to be silent."
        ],
        [
            "Be a voice, not an echo."
        ],
        [
            "I'm sorry Dave. I'm afraid I can't do that."
        ],
        [
            "This cake is great. It's so delicious and moist."
        ],
        [
            "Prior to November 22, 1963."
        ]
    ],
    "eval_split_max_size": null,
    "eval_split_size": 0.2,
    "model_args": {
        "num_chars": 66,
        "out_channels": 513,
        "spec_segment_size": 62,
        "hidden_channels": 192,
        "hidden_channels_ffn_text_encoder": 768,
        "num_heads_text_encoder": 2,
        "num_layers_text_encoder": 10,
        "kernel_size_text_encoder": 3,
        "dropout_p_text_encoder": 0.1,
        "dropout_p_duration_predictor": 0.5,
        "kernel_size_posterior_encoder": 5,
        "dilation_rate_posterior_encoder": 1,
        "num_layers_posterior_encoder": 16,
        "kernel_size_flow": 5,
        "dilation_rate_flow": 1,
        "num_layers_flow": 4,
        "resblock_type_decoder": "2",
        "resblock_kernel_sizes_decoder": [
            3,
            7,
            11
        ],
        "resblock_dilation_sizes_decoder": [
            [
                1,
                3,
                5
            ],
            [
                1,
                3,
                5
            ],
            [
                1,
                3,
                5
            ]
        ],
        "upsample_rates_decoder": [
            8,
            8,
            2,
            2
        ],
        "upsample_initial_channel_decoder": 512,
        "upsample_kernel_sizes_decoder": [
            16,
            16,
            4,
            4
        ],
        "use_sdp": true,
        "noise_scale": 1.0,
        "inference_noise_scale": 0.3,
        "length_scale": 1.5,
        "noise_scale_dp": 0.6,
        "inference_noise_scale_dp": 0.3,
        "max_inference_len": null,
        "init_discriminator": true,
        "use_spectral_norm_disriminator": false,
        "use_speaker_embedding": false,
        "num_speakers": 10,
        "speakers_file": "/home/ubuntu/Custom_TTS_train/Result/vits_approach2-April-08-2022_08+32AM-0000000/speakers.json",
        "d_vector_file": "/home/ubuntu/Custom_TTS_train/embeddings/speaker_embeddings.json",
        "speaker_embedding_channels": 512,
        "use_d_vector_file": true,
        "d_vector_dim": 512,
        "detach_dp_input": true,
        "use_language_embedding": false,
        "embedded_language_dim": 4,
        "num_languages": 1,
        "language_ids_file": null,
        "use_speaker_encoder_as_loss": false,
        "speaker_encoder_config_path": "train_tts_approach/config_se.json",
        "speaker_encoder_model_path": "train_tts_approach/model_se.pth.tar",
        "condition_dp_on_speaker": true,
        "freeze_encoder": false,
        "freeze_DP": false,
        "freeze_PE": false,
        "freeze_flow_decoder": false,
        "freeze_waveform_decoder": false
    },
    "lr_gen": 0.00002,
    "lr_disc": 0.00002,
    "lr_scheduler_gen": "ExponentialLR",
    "lr_scheduler_gen_params": {
        "gamma": 0.999875,
        "last_epoch": -1
    },
    "lr_scheduler_disc": "ExponentialLR",
    "lr_scheduler_disc_params": {
        "gamma": 0.999875,
        "last_epoch": -1
    },
    "kl_loss_alpha": 1.0,
    "disc_loss_alpha": 1.0,
    "gen_loss_alpha": 1.0,
    "feat_loss_alpha": 1.0,
    "mel_loss_alpha": 45.0,
    "dur_loss_alpha": 1.0,
    "speaker_encoder_loss_alpha": 1.0,
    "return_wav": true,
    "r": 1,
    "num_speakers": 10,
    "use_speaker_embedding": false,
    "speakers_file": "/home/ubuntu/Custom_TTS_train/Result/vits_approach2-April-08-2022_08+32AM-0000000/speakers.json",
    "speaker_embedding_channels": 256,
    "language_ids_file": null,
    "use_language_embedding": false,
    "use_d_vector_file": true,
    "d_vector_file": "/home/ubuntu/Custom_TTS_train/embeddings/speaker_embeddings.json",
    "d_vector_dim": 512
}
```

**Comment by OmiWakode** [comment]
Also, for the same config, with 7667 wav files (11 hours of data)  and training for 200,000 steps, avg_loss_1 is varying between 25-27. Audio quality isn't improving with constant static voice. Here is the [tensorboard link](http://13.234.66.149:5000/#audio). 
Any suggestions to reduce the loss of values and increase the quality of audio?

**Comment by p0p4k** [comment]
Hi, I am training my model on another language and got the same "None" error. I believe it is from "add_blank=True". Make it false and it will disappear.  (check TTSTokenizer class in tokenizer.py to understand how the blank works. It will use "blank" or "pad".)
As for the output quality, I am not sure what to do for that exactly. Maybe there is lot of noise in the training data? Maybe train a small subset of data (single speaker) and test the model (my model just worked with 10 min of data using a pre-trained model within 10k steps, 48 batch size! But, there is slight noise carry over from the original samples to the output).

**Comment by p0p4k** [comment]
Although, it wont make a difference, but you have two different values for "speaker_embedding_channels".
